### PR TITLE
Update StoryRecorder.java

### LINKED ISSUE #9602: Compiler crashes with -XsoycDetailed
**GWT version:** 2.8.2
**Browser (with version):** N/A
**Operating System:** Windows 8.1 

---

##### Description

When using the flags ` -compileReport -XcompilerMetrics -XsoycDetailed`, the 
compiler fails with `java.util.EmptyStackException`
`at com.google.gwt.core.ext.soyc.impl.StoryRecorder.analyzeFragment(StoryRecorder.java:179)`

The code responsible for this is at `StoryRecorder.java:179`:
```
        while (!dependencyScope.peek().range.contains(range)) {
          popAndRecord(dependencyScope, fragment);
        }
```
It's trying to `peek()` from an empty stack.  The solution might be as simple as adding a check for `isEmpty()`, like this:
```
        while (!dependencyScope.isEmpty() && !dependencyScope.peek().range.contains(range)) {
          popAndRecord(dependencyScope, fragment);
        }
```

##### Steps to reproduce

Do a compile with the flags ` -compileReport -XcompilerMetrics -XsoycDetailed`

##### Known workarounds

Don't use the `-XsoycDetailed` flag

##### Links to further discussions

[Stack Overflow question "Maven Compilation for GWT 2.7 SOYC Fails"](https://stackoverflow.com/questions/27897770/maven-compilation-for-gwt-2-7-soyc-fails)

### LINKED ISSUE #9635: StackEmptyException in StoryRecorder.java
Gwt 2.8.2 

##### Description
prevent StackEmptyException in StoryRecorder.java

[INFO] java.util.EmptyStackException
[INFO]  at java.util.Stack.peek(Stack.java:102)
[INFO]  at com.google.gwt.core.ext.soyc.impl.StoryRecorder.analyzeFragment(StoryRecorder.java:179)
[INFO]  at com.google.gwt.core.ext.soyc.impl.StoryRecorder.recordStoriesImpl(StoryRecorder.java:132)
[INFO]  at com.google.gwt.core.ext.soyc.impl.StoryRecorder.recordStories(StoryRecorder.java:73)
[INFO]  at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.makeSoycArtifacts(JavaToJavaScriptCompiler.java:843)
[INFO]  at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.addSoycArtifacts(JavaToJavaScriptCompiler.java:710)
[INFO]  at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.addSyntheticArtifacts(JavaToJavaScriptCompiler.java:731)
[INFO]  at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.compilePermutation(JavaToJavaScriptCompiler.java:454)
[INFO]  at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.compilePermutation(JavaToJavaScriptCompiler.java:272)
[INFO]  at com.google.gwt.dev.CompilePerms.compile(CompilePerms.java:198)
[INFO]  at com.google.gwt.dev.ThreadedPermutationWorkerFactory$ThreadedPermutationWorker.compile(ThreadedPermutationWorkerFactory.java:50)
[INFO]  at com.google.gwt.dev.PermutationWorkerFactory$Manager$WorkerThread.run(PermutationWorkerFactory.java:74)

##### Steps to reproduce

if nested contexts are empty? 
(project is proprietary unfortunately)

##### Links to further discussions

https://gitter.im/gwtproject/gwt?at=5bc5f887e65a634336ea02d3 

https://github.com/gwtproject/gwt/pull/9634/commits/99ddc2817e5ea498e2856276b3fbe8aeae40a40d

### LINKED ISSUE #9804: CompileReport exception
**GWT version:** 2.10.0
**Browser (with version):** 
**Operating System:** 

---

##### Description

EmptyStackException when compiling with -compileReport option:

 [java] Compiling module org.gwtsvg.edrawsvg
     [java]    Compiling 1 permutation
     [java]       Compiling permutation 0...
     [java]       Creating Stories file for the compile report
     [java]          [WARN] Could not write dependency file; proceeding anyway.
     [java] java.util.EmptyStackException
     [java]     at java.util.Stack.peek(Stack.java:102)
     [java]     at com.google.gwt.core.ext.soyc.impl.StoryRecorder.analyzeFragment(StoryRecorder.java:179)
     [java]     at com.google.gwt.core.ext.soyc.impl.StoryRecorder.recordStoriesImpl(StoryRecorder.java:132)
     [java]     at com.google.gwt.core.ext.soyc.impl.StoryRecorder.recordStories(StoryRecorder.java:73)
     [java]     at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.makeSoycArtifacts(JavaToJavaScriptCompiler.java:843)
     [java]     at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.addSoycArtifacts(JavaToJavaScriptCompiler.java:710)
     [java]     at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.addSyntheticArtifacts(JavaToJavaScriptCompiler.java:731)
     [java]     at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.compilePermutation(JavaToJavaScriptCompiler.java:454)
     [java]     at com.google.gwt.dev.jjs.JavaToJavaScriptCompiler.compilePermutation(JavaToJavaScriptCompiler.java:272)
     [java]     at com.google.gwt.dev.CompilePerms.compile(CompilePerms.java:198)
     [java]     at com.google.gwt.dev.ThreadedPermutationWorkerFactory$ThreadedPermutationWorker.compile(ThreadedPermutationWorkerFactory.java:50)
     [java]     at com.google.gwt.dev.PermutationWorkerFactory$Manager$WorkerThread.run(PermutationWorkerFactory.java:74)
     [java]     at java.lang.Thread.run(Thread.java:748)
     [java]    Compile of permutations succeeded
     [java]    Compilation succeeded -- 275,544s

**The loop :**
while (!dependencyScope.peek().range.contains(range)) {
          popAndRecord(dependencyScope, fragment);
        }
**Should be:**
while (!dependencyScope.isEmpty() && !dependencyScope.peek().range.contains(range)) {
          popAndRecord(dependencyScope, fragment);
        }

### PR DESCRIPTION
**Pull request by DRAWSVG**
Fixes https://github.com/gwtproject/gwt/issues/9635
Fixes https://github.com/gwtproject/gwt/issues/9602
Fixes https://github.com/gwtproject/gwt/issues/9804

### CODE REVIEW

**Comment by zbynek** [comment]
@DRAWSVG can you please update the initial comment of the PR  -- remove th default text and include the relevant links e.g. 
Fixes #9635
Fixes #9602 
Fixes #9804

File: dev/core/src/com/google/gwt/core/ext/soyc/impl/StoryRecorder.java [hunk 1]
```diff
@@ -351,7 +348,9 @@ private void recordStory(SourceInfo info, int fragment, int length, Range range)
       theStory = new StoryImpl(storyCache.get(info), length);
     }
 
-    emitStory(theStory, range);
+    if (range.getStart() < js[curHighestFragment].length()) {
```
Comment (zbynek) [inline dev/core/src/com/google/gwt/core/ext/soyc/impl/StoryRecorder.java:351]: the check `start < length` was added outside of `emitStory`, check that `end <= length` was added inside  that method -- maybe both checks should be in one place -- inside `emitStory`.

**Comment by zbynek** [comment]
With the patch applied I was able to verify that there are no more exceptions in console and `detailedStories0.xml.gz`  is a complete XML file (before it was cut short because of the exception).

**Comment by niloc132** [comment]
@DRAWSVG This hasn't been merged yet, I closed the other as it was incomplete compared to this. This pull request is waiting for your reply or update in response to zbynek's comment at https://github.com/gwtproject/gwt/pull/9808#discussion_r1140465818.

Unless you are abandoning this change, please reopen it.

**Comment by niloc132** [comment]
Please take a look at https://github.com/gwtproject/gwt/pull/9808#discussion_r1140465818 to see if it makes sense to either make the suggested change, or reply to explain why we should skip that.

File: dev/core/src/com/google/gwt/core/ext/soyc/impl/StoryRecorder.java (hunk 1 above)
Comment (DRAWSVG) [reply dev/core/src/com/google/gwt/core/ext/soyc/impl/StoryRecorder.java:351]: I do not think so.
It seems that the value of the end of range property is wrong.
The call to emitStory must be kept to get the description of the story.

Comment (zbynek) [reply dev/core/src/com/google/gwt/core/ext/soyc/impl/StoryRecorder.java:351]: OK, feel free to disregard

**Pull request merged**